### ISSUE TITLE
Some partners question some precious problems, and we'd like to display them here.

### ISSUE THREAD

**Issue by idealwei** [starmemda/CAMoE#4]
Hi, I'm trying to evaluate the effectiveness of DSL loss. I've tested it with CLIP4Clip, but the improvement was not consistent with you paper. I got 44.1%@V2T-R@1 for original loss and 44.6%@V2T-R@1 for DSL loss.  I revised simlarity matrix both on training and infernece time. Can you help me ?

**Comment by starmemda** [comment]
The temperature hyperparameter is used for smoothing the gradient in training and should be set to 1 during inference.

**Comment by starmemda** [comment]
Waiting for your result.

**Comment by idealwei** [comment]
> The temperature hyperparameter is used for smoothing the gradient in training and should be set to 1 during inference.

Thank you for your reply. I'll  have a try.  
I'm instereted in multi-modal-learning area and welcome to meet someone work on this area as well. Can I have your personal concact for further discussion ?

**Comment by starmemda** [comment]
You can contact me by emails:  sttarmemeda@gmail.com.

**Comment by idealwei** [comment]
😄

**Comment by idealwei** [comment]
Hi，starmemda：
Really doubt the effectiveness of DSL loss ！Please check it thoroughly.
First, with your suggestion, I got 49.0% V2T-R@1 and 48.3% T2V-R@1 on MSR-VTT with DSL loss, which seemed to be a successful attempt.
However, I tried to train CLIP4Clip with original loss and evaluate with the "revised sim_matrix" referred in your paper. 
The result  is also extremely good (49.0% V2T-R@1 and 47.8% T2V-R@1 on MSR-VTT).
So We can conclude  that R@K improvements was brought by the "revised sim_matrix" rather than the DSL loss.
Furthermore, Using the "revised sim_matrix" in evaluation  seems not fair. 
Waiting for your reply !

**Comment by starmemda** [comment]
Thanks for your advice, we also find that utilizing dual softmax for inference improves a lot.  It's a wonderful discovery that the metric learning method solving content heterogeneity in VTR works so much.  
Maybe we should write the paper from the standing point of proposing a novel metric learning method just like  K-Reciprocal Encoding.  But employing dual softmax to train indeed improves more. Considering all these situations, it may be more appropriate to train with DSL, which conforms to our paper.
Later we will perfect the paper and codes to alleviate the misunderstandings.

**Comment by Stephen0808** [comment]
Maybe release the model(CAMoE) source code could help the readers reproduce the result in the paper.

**Comment by starmemda** [comment]
We'd like to, but not now.

**Comment by zimenglan-sysu-512** [comment]
> The temperature hyperparameter is used for smoothing the gradient in training and should be set to 1 during inference.

hi @starmemda it means that temperature=1000 in training, but temperature=1 in inference stage?

**Comment by Lemonahahah** [comment]
The use of DSL in testing seems to be an implicit data leak, which violates the consensus in this field. Unfortunately, this leak seems to be the focus of work and the source of improvement.

**Comment by starmemda** [comment]
> The use of DSL in testing seems to be an implicit data leak, which violates the consensus in this field. Unfortunately, this leak seems to be the focus of work and the source of improvement.

It seems you mistake DSL. It doesn't refer to data's order information. So your concern maybe not true.

**Comment by Lemonahahah** [comment]
> > The use of DSL in testing seems to be an implicit data leak, which violates the consensus in this field. Unfortunately, this leak seems to be the focus of work and the source of improvement.
> 
> It seems you mistake DSL. It doesn't refer to data's order information. So your concern maybe not true.

The premise of DSL is that we know the one-to-one correspondence between the video and the text, and then we can use the retrieval probabilities of the two directions to re-score, right? During training, this setting is ok. But when testing, it should be something wrong.

**Comment by starmemda** [comment]
> > > The use of DSL in testing seems to be an implicit data leak, which violates the consensus in this field. Unfortunately, this leak seems to be the focus of work and the source of improvement.
> > 
> > 
> > It seems you mistake DSL. It doesn't refer to data's order information. So your concern maybe not true.
> 
> The premise of DSL is that we know the one-to-one correspondence between the video and the text, and then we can use the retrieval probabilities of the two directions to re-score, right? During training, this setting is ok. But when testing, it should be something wrong.

But DSL can improve even more on the multi-sentence eval dataset. The problem it solves truly is the over-generalized cases.

**Comment by starmemda** [comment]
> 

We think it provides a wonderful direction that focusing on the data distribution may be one of the future key methods.

**Comment by Lemonahahah** [comment]
> > > > The use of DSL in testing seems to be an implicit data leak, which violates the consensus in this field. Unfortunately, this leak seems to be the focus of work and the source of improvement.
> > > 
> > > 
> > > It seems you mistake DSL. It doesn't refer to data's order information. So your concern maybe not true.
> > 
> > 
> > The premise of DSL is that we know the one-to-one correspondence between the video and the text, and then we can use the retrieval probabilities of the two directions to re-score, right? During training, this setting is ok. But when testing, it should be something wrong.
> 
> But DSL can improve even more on the multi-sentence eval dataset. The problem it solves truly is the over-generalized cases.

There is only one important thing. Is the DSL done after obtaining the corresponding relationship between the video and the text? My current judgment is, yes, the premise of DSL work is to know the correspondence between them. So, I understand that this is an improvement obtained by using data leakage.

**Comment by starmemda** [comment]
> There is only one important thing. Is the DSL done after obtaining the corresponding relationship between the video and the text? My current judgment is, yes, the premise of DSL work is to know the correspondence between them. So, I understand that this is an improvement obtained by using data leakage.

No, DSL doesn't obtain any corresponding relationship in advance. The only assumption is that every query can be matched with some gallery keys, which is the same as all previous methods. And some newest methods are taking DSL as post-processing means.

**Comment by Lemonahahah** [comment]
> > There is only one important thing. Is the DSL done after obtaining the corresponding relationship between the video and the text? My current judgment is, yes, the premise of DSL work is to know the correspondence between them. So, I understand that this is an improvement obtained by using data leakage.
> 
> No, DSL doesn't obtain any corresponding relationship in advance. The only assumption is that every query can be matched with some gallery keys, which is the same as all previous methods. And some newest methods are taking DSL as post-processing means.

Your reply made me doubt my judgment. I would like to ask a question. If the model does not know the correspondence between the video and the text, how can the two matrices be added together bitwise in testing?

**Comment by starmemda** [comment]
> Your reply made me doubt my judgment. I would like to ask a question. If the model does not know the correspondence between the video and the text, how can the two matrices be added together bitwise in testing?

The Hadamard product presented in this paper is independent of the order in which the specific query appears. So it doesn't obtain the correspondence you mentioned at all.

**Comment by starmemda** [comment]
> > > There is only one important thing. Is the DSL done after obtaining the corresponding relationship between the video and the text? My current judgment is, yes, the premise of DSL work is to know the correspondence between them. So, I understand that this is an improvement obtained by using data leakage.
> > 
> > 
> > No, DSL doesn't obtain any corresponding relationship in advance. The only assumption is that every query can be matched with some gallery keys, which is the same as all previous methods. And some newest methods are taking DSL as post-processing means.
> 
> Your reply made me doubt my judgment. I would like to ask a question. If the model does not know the correspondence between the video and the text, how can the two matrices be added together bitwise in testing?

**Comment by Lemonahahah** [comment]
> > Your reply made me doubt my judgment. I would like to ask a question. If the model does not know the correspondence between the video and the text, how can the two matrices be added together bitwise in testing?
> 
> The Hadamard product presented in this paper is independent of the order in which the specific query appears. So it doesn't obtain the correspondence you mentioned at all.

![image](https://user-images.githubusercontent.com/94953627/144243518-a8707433-ba79-4cb9-a1ef-2f2ed18a61f3.png)
This should be a streamlined version of the DSL. 

The following picture show the process:
![image](https://user-images.githubusercontent.com/94953627/144244346-15048b0f-5ac1-4b91-a76f-2df87bfac2ed.png)
I think this understanding is okay, right?

**Comment by Lemonahahah** [comment]
> > > Your reply made me doubt my judgment. I would like to ask a question. If the model does not know the correspondence between the video and the text, how can the two matrices be added together bitwise in testing?
> > 
> > 
> > The Hadamard product presented in this paper is independent of the order in which the specific query appears. So it doesn't obtain the correspondence you mentioned at all.
> 
> ![image](https://user-images.githubusercontent.com/94953627/144243518-a8707433-ba79-4cb9-a1ef-2f2ed18a61f3.png) This should be a streamlined version of the DSL.
> 
> The following picture show the process: ![image](https://user-images.githubusercontent.com/94953627/144244346-15048b0f-5ac1-4b91-a76f-2df87bfac2ed.png) I think this understanding is okay, right?

The premise of the above operation is that video-to-text and text-to-video with the same index are in the same matrix position. In other words, the calculation of DSL depends on this matrix (V2T and T2V). The premise that their positions correspond correctly is that you know the true alignment relationship between the video and the text. This is a data leakage I mentioned above.

**Comment by starmemda** [comment]
The relevant symbol you mentioned is somewhat fuzzy and I can't fully understand.  And maybe I should emphasize that the similarity matrix and probability matrix in the paper are different.

**Comment by Lemonahahah** [comment]
> The relevant symbol you mentioned is somewhat fuzzy and I can't fully understand. And maybe I should emphasize that the similarity matrix and probability matrix in the paper are different.

Let me confirm one thing, can the calculation in this figure be understood as the main calculation process of DSL?
![image](https://user-images.githubusercontent.com/94953627/144249458-67e8734e-64f4-432a-914c-4cd15bd9c938.png)

**Comment by starmemda** [comment]
> > The relevant symbol you mentioned is somewhat fuzzy and I can't fully understand. And maybe I should emphasize that the similarity matrix and probability matrix in the paper are different.
> 
> Let me confirm one thing, can the calculation in this figure be understood as the main calculation process of DSL? ![image](https://user-images.githubusercontent.com/94953627/144249458-67e8734e-64f4-432a-914c-4cd15bd9c938.png)

No, it is quite different from what we do.

**Comment by Lemonahahah** [comment]
> > > The relevant symbol you mentioned is somewhat fuzzy and I can't fully understand. And maybe I should emphasize that the similarity matrix and probability matrix in the paper are different.
> > 
> > 
> > Let me confirm one thing, can the calculation in this figure be understood as the main calculation process of DSL? ![image](https://user-images.githubusercontent.com/94953627/144249458-67e8734e-64f4-432a-914c-4cd15bd9c938.png)
> 
> No, it is quite different from what we do.

So how is DSL used during testing?

**Comment by starmemda** [comment]
> > > > The relevant symbol you mentioned is somewhat fuzzy and I can't fully understand. And maybe I should emphasize that the similarity matrix and probability matrix in the paper are different.
> > > 
> > > 
> > > Let me confirm one thing, can the calculation in this figure be understood as the main calculation process of DSL? ![image](https://user-images.githubusercontent.com/94953627/144249458-67e8734e-64f4-432a-914c-4cd15bd9c938.png)
> > 
> > 
> > No, it is quite different from what we do.
> 
> So how is DSL used during testing?

We'd like to supplement it soon.

**Comment by Lemonahahah** [comment]
> > > > > The relevant symbol you mentioned is somewhat fuzzy and I can't fully understand. And maybe I should emphasize that the similarity matrix and probability matrix in the paper are different.
> > > > 
> > > > 
> > > > Let me confirm one thing, can the calculation in this figure be understood as the main calculation process of DSL? ![image](https://user-images.githubusercontent.com/94953627/144249458-67e8734e-64f4-432a-914c-4cd15bd9c938.png)
> > > 
> > > 
> > > No, it is quite different from what we do.
> > 
> > 
> > So how is DSL used during testing?
> 
> We'd like to supplement it soon.

Fine.

**Comment by Lemonahahah** [comment]
![image](https://user-images.githubusercontent.com/94953627/144258099-282a0064-b6ef-4683-bf38-a309e16abe04.png)
![image](https://user-images.githubusercontent.com/94953627/144258763-a227b0c7-9c03-442b-a21d-1089b91e8035.png)

I found the process in the above figure(in your paper)  seem to be the same as the below.

![image](https://user-images.githubusercontent.com/94953627/144258224-38fe8631-ded1-411a-b346-4faf4715b10f.png)

**Comment by EmiliaKKK** [comment]
> ![image](https://user-images.githubusercontent.com/94953627/144258099-282a0064-b6ef-4683-bf38-a309e16abe04.png) ![image](https://user-images.githubusercontent.com/94953627/144258763-a227b0c7-9c03-442b-a21d-1089b91e8035.png)
> 
> I found the process in the above figure(in your paper) seem to be the same as the below.
> 
> ![image](https://user-images.githubusercontent.com/94953627/144258224-38fe8631-ded1-411a-b346-4faf4715b10f.png)

![MommyTalk1638769684607](https://user-images.githubusercontent.com/57026602/144794810-67a6e531-0b9c-4007-b31f-2e00cea481dc.jpg)
I think the equation above is more likely to be the true calculation in the paper. 
![](https://user-images.githubusercontent.com/94953627/144258224-38fe8631-ded1-411a-b346-4faf4715b10f.png)
The second equation is an revised one used in CLIP2TV for inference only.
I think they are not the same for the frist equation may have some consideration on gradient smoothing.
But from the prospective of data leak, maybe they are the same. I'm in fact confused about the "data leak". Could you please explain it in detail? In my view, DSL just introduces some T2V information to V2T task.

**Comment by Lemonahahah** [comment]
After my thinking, there should be no leaks here. But I still think this approach is weird. In fact, this method can not be used for the actual video-text retrieval system, but is valuable for this data set evaluation method.

**Comment by Lemonahahah** [comment]
> > ![image](https://user-images.githubusercontent.com/94953627/144258099-282a0064-b6ef-4683-bf38-a309e16abe04.png) ![image](https://user-images.githubusercontent.com/94953627/144258763-a227b0c7-9c03-442b-a21d-1089b91e8035.png)
> > I found the process in the above figure(in your paper) seem to be the same as the below.
> > ![image](https://user-images.githubusercontent.com/94953627/144258224-38fe8631-ded1-411a-b346-4faf4715b10f.png)
> 
> ![MommyTalk1638769684607](https://user-images.githubusercontent.com/57026602/144794810-67a6e531-0b9c-4007-b31f-2e00cea481dc.jpg) I think the equation above is more likely to be the true calculation in the paper. ![](https://user-images.githubusercontent.com/94953627/144258224-38fe8631-ded1-411a-b346-4faf4715b10f.png) The second equation is an revised one used in CLIP2TV for inference only. I think they are not the same for the frist equation may have some consideration on gradient smoothing. But from the prospective of data leak, maybe they are the same. I'm in fact confused about the "data leak". Could you please explain it in detail? In my view, DSL just introduces some T2V information to V2T task.

After my thinking, there should be no leaks here. But I still think this approach is weird. In fact, this method can not be used for the actual video-text retrieval system, but is valuable for this data set evaluation method.

**Comment by Lemonahahah** [comment]
> > > > > The relevant symbol you mentioned is somewhat fuzzy and I can't fully understand. And maybe I should emphasize that the similarity matrix and probability matrix in the paper are different.
> > > > 
> > > > 
> > > > Let me confirm one thing, can the calculation in this figure be understood as the main calculation process of DSL? ![image](https://user-images.githubusercontent.com/94953627/144249458-67e8734e-64f4-432a-914c-4cd15bd9c938.png)
> > > 
> > > 
> > > No, it is quite different from what we do.
> > 
> > 
> > So how is DSL used during testing?
> 
> We'd like to supplement it soon.

I have figured out the process, thank you for your answers, there is no problem with this method.

**Comment by starmemda** [comment]
> > > ![image](https://user-images.githubusercontent.com/94953627/144258099-282a0064-b6ef-4683-bf38-a309e16abe04.png) ![image](https://user-images.githubusercontent.com/94953627/144258763-a227b0c7-9c03-442b-a21d-1089b91e8035.png)
> > > I found the process in the above figure(in your paper) seem to be the same as the below.
> > > ![image](https://user-images.githubusercontent.com/94953627/144258224-38fe8631-ded1-411a-b346-4faf4715b10f.png)
> > 
> > 
> > ![MommyTalk1638769684607](https://user-images.githubusercontent.com/57026602/144794810-67a6e531-0b9c-4007-b31f-2e00cea481dc.jpg) I think the equation above is more likely to be the true calculation in the paper. ![](https://user-images.githubusercontent.com/94953627/144258224-38fe8631-ded1-411a-b346-4faf4715b10f.png) The second equation is an revised one used in CLIP2TV for inference only. I think they are not the same for the frist equation may have some consideration on gradient smoothing. But from the prospective of data leak, maybe they are the same. I'm in fact confused about the "data leak". Could you please explain it in detail? In my view, DSL just introduces some T2V information to V2T task.
> 
> After my thinking, there should be no leaks here. But I still think this approach is weird. In fact, this method can not be used for the actual video-text retrieval system, but is valuable for this data set evaluation method.

**Comment by bnu-wangxun** [comment]
> 
What is the value of temp used in training?

**Comment by SCZwangxiao** [comment]
> > > ![image](https://user-images.githubusercontent.com/94953627/144258099-282a0064-b6ef-4683-bf38-a309e16abe04.png) ![image](https://user-images.githubusercontent.com/94953627/144258763-a227b0c7-9c03-442b-a21d-1089b91e8035.png)
> > > I found the process in the above figure(in your paper) seem to be the same as the below.
> > > ![image](https://user-images.githubusercontent.com/94953627/144258224-38fe8631-ded1-411a-b346-4faf4715b10f.png)
> > 
> > 
> > ![MommyTalk1638769684607](https://user-images.githubusercontent.com/57026602/144794810-67a6e531-0b9c-4007-b31f-2e00cea481dc.jpg) I think the equation above is more likely to be the true calculation in the paper. ![](https://user-images.githubusercontent.com/94953627/144258224-38fe8631-ded1-411a-b346-4faf4715b10f.png) The second equation is an revised one used in CLIP2TV for inference only. I think they are not the same for the frist equation may have some consideration on gradient smoothing. But from the prospective of data leak, maybe they are the same. I'm in fact confused about the "data leak". Could you please explain it in detail? In my view, DSL just introduces some T2V information to V2T task.
> 
> After my thinking, there should be no leaks here. But I still think this approach is weird. In fact, this method can not be used for the actual video-text retrieval system, but is valuable for this data set evaluation method.

I think this method could be used in the actual video-text retrieval system, but only in limited scenarios. In a general search scenario, the method is not practical. Because we only have one query and many candidate items, and there is no video to text information.
